crowbar_framework/lib/crowbar/deployment_queue.rb

```diff
@@ -0,0 +1,410 @@
+            delay, pre_cached_nodes = elements_not_ready(nodes_map.keys)
+            list << item if delay.empty?
+          end
+
+          # Update the queue. Drop all proposals that we can process now (list) and those
+          # that are deleted (remove_list). This leaves in the queue only proposals
+          # which are still waiting for nodes (delay not empty), or for which deps are not
+          # ready/created/deployed (queue_me = true).
+          save_db = false
+          remove_list.each do |iii|
+            save_db |= dequeue_proposal_no_lock(db["proposal_queue"], iii["inst"], iii["barclamp"])
+          end
+
+          list.each do |iii|
+            save_db |= dequeue_proposal_no_lock(db["proposal_queue"], iii["inst"], iii["barclamp"])
+          end
+
+          db.save if save_db
+
+        rescue StandardError => e
+          logger.error("Error processing queue: #{e.message}")
+          logger.debug("process queue: exit: error")
+          return
+        ensure
+          file_lock.release(f)
+        end
+
+        logger.debug("process queue: list: #{list.inspect}")
+
+        # For each ready item, apply it.
+        list.each do |item|
+          logger.debug("process queue: item to do: #{item.inspect}")
+
+          bc = item["barclamp"]
+          inst = item["inst"]
+          service = eval("#{bc.camelize}Service.new logger")
+
+          # This will call apply_role and chef-client.
+          # Params: (inst, in_queue, validate_after_save)
+          answer = service.proposal_commit(inst, true, false)
+
+          logger.debug("process queue: item #{item.inspect}: results #{answer.inspect}")
+
+          # 202 means some nodes are not ready, bail out in that case
+          # We're re-running the whole apply continuously, until there
+          # are no items left in the queue.
+          # FIXME: This is lame, because from the user perspective, we're still
+          # applying the first barclamp, while this part was in fact already
+          # completed and we're applying next item(s) in the queue.
+          loop_again = true if answer[0] != 202
+
+          $htdigest_reload = true
+        end
+        logger.debug("process queue: exit")
+      end
+    end
+
+    private
+
+    def file_lock
+      FileLock
+    end
+
+    # Removes the proposal reference from the queue, updates the proposal as not queued
+    # and drops the 'pending roles' from the affected nodes.
+    def dequeue_proposal_no_lock(bc, inst, queue)
+      logger.debug("dequeue_proposal_no_lock: enter #{inst} #{bc}")
+      begin
+        elements = nil
+
+        # The elements = item["elements"] is on purpose to get the assignment out of the element.
+        # Find the proposal to delete, get its elements (nodes)
+        queue.delete_if { |item| item["barclamp"] == bc and item["inst"] == inst and ((elements = item["elements"]) or true)}
```

Comment:
Style/AndOr: Use `&&` instead of `and`.<br>Style/AndOr: Use `||` instead of `or`.<br>Style/SpaceInsideBlockBraces: Space missing inside }.